Library/Homebrew/dependency.rb

```diff
@@ -124,11 +124,39 @@ def merge_repeats(all)
       all.map(&:name).uniq.map do |name|
         deps = grouped.fetch(name)
         dep  = deps.first
-        tags = deps.flat_map(&:tags).uniq
+        tags = merge_tags(deps)
         option_names = deps.flat_map(&:option_names).uniq
         dep.class.new(name, tags, dep.env_proc, option_names)
       end
     end
+
+    private
+
+    def merge_tags(deps)
+      options = deps.flat_map(&:option_tags).uniq
+      merge_necessity(deps) + merge_temporality(deps) + options
+    end
+
+    def merge_necessity(deps)
+      if deps.any? { |dep| !dep.recommended? && !dep.optional? }
+        # FIXME: Should be `required?`, but `required?` implies `!build?`.
+        [] # Implies required dependency.
+      elsif deps.any?(&:recommended?)
+        [:recommended]
+      else # deps.all?(&:optional?)
+        [:optional]
+      end
+    end
+
+    def merge_temporality(deps)
+      if deps.all?(&:build?)
+        [:build]
+      elsif deps.all?(&:run?)
+        [:run]
+      else
+        [] # Implies build & runtime dependency.
```

Comment:
Mind using `and` instead of `&` just to avoid confusion with the actual `&` operator?